### PR TITLE
Fail closed when a login resolves no identity

### LINKED ISSUE #95: OID/SAML: fail closed when a login is valid but resolves no username
## Summary

An OpenID login can end up **valid with a null/empty username**, which then throws downstream (a 500) instead of taking the clean invalid-login path. Surfaced by the Copilot reviewer on #94 (a behavior-preserving refactor); the behavior is **pre-existing on `main`**, so it is tracked here rather than changed inside that no-op extraction.

## How it happens

In `OidPost` (now via `OidcAuthorizeStateBuilder.Build`), login validity and the username are derived independently:

- `Valid` becomes true if a role matches the allow-list (`grants.Valid`, i.e. `config.Roles` is non-empty and a role claim matches), **or** if the preferred-username/`DefaultUsernameClaim` claim is present while no allow-list is configured.
- `Username` is set **only** by the preferred-username / `DefaultUsernameClaim` branch, or by the `sub` fallback, and the `sub` fallback runs **only when `!Valid`**.

So if the provider grants validity purely through a **role allow-list match** but sends **no** preferred-username / `DefaultUsernameClaim` claim, `Valid == true` while `Username == null`, and the `sub` fallback is skipped (because `Valid` is already true).

## Impact

`OidAuth` then calls `CreateCanonicalLinkAndUserIfNotExist("oid", provider, timedState.Username, …)` with a null name; the create path reaches `_userManager.CreateUserAsync(null)`, which throws → **HTTP 500** instead of a clean invalid-login response. This is **fail-closed** (no unauthorized access is granted, the request errors out), but it is a poor failure mode and a latent correctness bug. The SAML path (`SamlAuth` → `GetNameID()`) should be checked for the analogous null/empty case.

## Proposed fix (separate PR)

In the callback, after deriving the authorize state, treat `Valid && string.IsNullOrEmpty(Username)` as an **invalid login** (return the existing invalid-login path, not a 500). Add negative tests: role-only allow-list validity with no username claim, and the SAML null-NameID analogue. Keep the change fail-closed.

Refs #89

### PR DESCRIPTION
**Pull request by iderex**
## What

An OpenID login could end up **valid with no resolved username** (role allow-list match while the provider sent no username claim; the `sub` fallback only runs when the login is not yet valid), and a SAML assertion could arrive **without a NameID**. Both crashed downstream on the null identity, surfacing as 500s instead of a clean denial, and the SAML linking endpoint could persist a dead empty-string link. `Closes #95`.

## How

- **OIDC**: the builder clamps validity to a resolved identity (`valid && !IsNullOrWhiteSpace(username)`). Every state consumer gates on `Valid` via `IsRedeemableBy`, so session minting and OID linking are both covered at the single producer.
- **SAML**: `Response.GetNameID` returns null instead of throwing on a missing NameID; `SamlAuth` rejects a null/empty/whitespace NameID as an invalid login (401, logged) at the same point in the guard order as the old crash (replay consumption unchanged).
- **Linking choke point**: `CreateCanonicalLink` refuses an unresolved identity (400) before the config mutation, no null-key crash, no dead `""` link persisted.
- **Belt**: `CreateCanonicalLinkAndUserIfNotExist` throws (mapped to the existing generic 403) should a future caller forget the guard.
- The OID denial log names both causes instead of misattributing everything to role claims, with the username sanitized inline.

## Why no working login regresses

Verified against Jellyfin 10.11's `UserManager` semantics: every newly-rejected input class previously produced a 500 (`GetUserByName`/`CreateUserAsync` reject null/empty/whitespace names; null dictionary keys throw). The one removed "working" flow only ever produced empty-keyed links no login could redeem, closing the identity-collapse channel where every empty-identity user mapped onto one account. Whitespace-only counts as unresolved for the same reason.

## Verification

- Tests first: the new negative tests were red on main (verified), 221/221 green now on a clean rebuild; six new tests pin the clamp (no-username / only-sub / empty / whitespace), the null NameID return, and the deliberate non-adoption of `sub` for role-valid logins.
- Full four-lens adversarial gate: bypass, correctness, and availability PASS (including a red-on-main probe and an empirical check of Jellyfin's username validation); integration returned NEEDS_IMPROVEMENT with two findings, the unguarded SAML linking path and the misattributed denial log, both fixed in this PR and confirmed by a focused re-verify (PASS).
- Three new real-server E2E checklist items cover the rejection-path behavior CI cannot exercise.

Closes #95

### CODE REVIEW

File: SSO-Auth/Api/SSOController.cs [hunk 1]
```diff
@@ -176,8 +176,8 @@ public async Task<ActionResult> OidPost(
             else
             {
                 _logger.LogWarning(
-                    "OpenID user {Username} has one or more incorrect role claims: {@Claims}. Expected any one of: {@ExpectedClaims}",
-                    timedState.Username,
+                    "OpenID login denied for {Username}: no role matched the allow-list, or the login resolved no username. Claims: {@Claims}. Roles expected (any one of): {@ExpectedClaims}",
+                    timedState.Username?.ReplaceLineEndings(string.Empty),
                     result.User.Claims.Select(o => new { o.Type, o.Value }),
                     config.Roles);
```
[the remark this answers is not in the document]
Comment (iderex) [reply SSO-Auth/Api/SSOController.cs:182]: Accepted and fixed in c336b02 - every provider-sourced value in the projection is now stripped inline (`Type`/`Value` in the OIDC denial log, and the same pattern in the SAML ACS denial log, which predated this PR but violates the same invariant). On the PII point: the claim values stay in the log deliberately - this Warning line is the only diagnostic an admin has for an allow-list/claim misconfiguration, and it fires only on denied logins; dropping the values would make the misconfiguration undebuggable. Documented trade-off.

File: SSO-Auth.Tests/SamlResponseTests.cs [hunk 2]
```diff
@@ -142,6 +142,15 @@ public void GetNameID_ValidResponse_ReturnsSignedSubject()
...
 
+    [Fact]
+    public void GetNameID_MissingNameId_ReturnsNull()
+    {
+        // An assertion without a NameID must yield null (the callback rejects it as an invalid
+        // login), not throw — previously this was an unhandled NRE turning into a 500 (#95).
+        var fixture = SamlTestFactory.Create(includeNameId: false);
+        Assert.Null(Load(fixture).GetNameID());
+    }
+
```
[the remark this answers is not in the document]
Comment (iderex) [reply SSO-Auth.Tests/SamlResponseTests.cs:153]: Agreed on the goal, deferring the implementation: there is no controller-level test infrastructure yet - SamlAuth needs SSOPlugin.Instance plus IUserManager/ISessionManager/ICryptoProvider doubles, none of which the suite has (it deliberately tests pure helpers; making the controller testable is the standing decomposition program). The reject path IS covered by three other layers in this PR: the pure-helper negative tests (null NameID return, validity clamp), the four-lens adversarial gate that traced the 401 path and its non-mutation empirically, and a new real-server E2E checklist item ("IdP omitting NameID → 401, no session, no user") that gates releases. A controller-level negative test lands when the session-minting step is extracted into a testable seam - tracked in the decomposition backlog. Not blocking this fix on building that infrastructure first.

**Review by iderex — COMMENTED** [review]
pre-merge review (iderex), walking the reviews item by item.

**Adversarial gate (full four-lens, behavior-changing login-path diff):** bypass, correctness, and availability PASS — including a red-on-main probe (the six new negative tests fail on main's code, pass here), an empirical check of Jellyfin 10.11's UserManager semantics proving no working login changes outcome (every newly-rejected class was a 500), and refutation of clamp evasion, sub-identity adoption, guard reordering, and fail-open on the null NameID path. Integration returned NEEDS_IMPROVEMENT with two findings — the SAML linking path outside the documented invariant and the misattributed denial log — both fixed (linking choke-point guard in CreateCanonicalLink; reworded, sanitized log) and confirmed by a focused delta re-verify (PASS), which also proved the OidLink invariant (the 400 guard is unreachable for a redeemable OID state).

**CodeRabbit, comment 1 (controller-path negative test):** goal agreed, implementation deferred with reasons on the thread — no controller test infrastructure exists (needs SSOPlugin.Instance + three service doubles); the reject path is covered by the pure-helper negative tests, the lens gate's empirical trace, and a new real-server E2E checklist item; a controller-level test lands when the session-minting seam is extracted (decomposition backlog).
**CodeRabbit, comment 2 ({@Claims} log projection unsanitized):** confirmed and fixed in c336b02 — every provider-sourced value is now stripped inline at both denial-log calls (including the pre-existing SAML ACS one); the values stay in the log deliberately as the only misconfiguration diagnostic, documented on the thread. Copilot: quota-blocked, no comments.

**Gates:** tests-first (new tests verified red on main), 221/221 green on a clean rebuild, build clean under `--warnaserror`, all functional CI checks green.
**Quality gate:** red on `new_coverage` ONLY (10.5% — the controller-side guards and log lines are not unit-testable today; the builder clamp is fully covered; verified via `api/qualitygates/project_status`, sole failing condition). Gate red on coverage only — accepted per free-plan policy; merging via the ruleset admin bypass.

**Overall verdict: APPROVE (sign-off).** Merging.
